// ark-tools/lib/reducers/filterReducer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var tableReducer = function tableReducer(_ref) {
  var key = _ref.key,
      _ref$filterState = _ref.filterState,
      filterState = _ref$filterState === void 0 ? false : _ref$filterState,
      _ref$filters = _ref.filters,
      filters = _ref$filters === void 0 ? [] : _ref$filters,
      _ref$filterOrig = _ref.filterOrig,
      filterOrig = _ref$filterOrig === void 0 ? [] : _ref$filterOrig;
  var SET_FILTERS = "".concat(key, "/SET_FILTERS");
  var OPEN_FILTER = "".concat(key, "/OPEN_FILTER");
  var CLOSE_FILTER = "".concat(key, "/CLOSE_FILTER");
  var CLEAR_FILTER = "".concat(key, "/CLEAR_FILTER");
  var APPLY_FILTER = "".concat(key, "/APPLY_FILTER");
  var CLEAR_ALL_SETTINGS = "".concat(key, "/CLEAR_ALL_SETTINGS");
  var tablesOptionsInitialState = {
    filterState: filterState,
    filters: filters,
    filterOrig: filterOrig
  };

  var tableOptionsReducer = function tableOptionsReducer() {
    var state = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : tablesOptionsInitialState;
    var action = arguments.length > 1 ? arguments[1] : undefined;

    switch (action.type) {
      case SET_FILTERS:
        return _objectSpread({}, state, {
          filters: action.filters
        });

      case OPEN_FILTER:
        return _objectSpread({}, state, {
          filterState: true,
          filterOrig: state.filters
        });

      case CLOSE_FILTER:
        return _objectSpread({}, state, {
          filterState: false,
          filters: state.filterOrig
        });

      case CLEAR_FILTER:
        return _objectSpread({}, state, {
          filterState: tablesOptionsInitialState.filterState,
          filters: tablesOptionsInitialState.filters
        });

      case APPLY_FILTER:
        return _objectSpread({}, state, {
          filterState: false,
          filterOrig: state.filters
        });

      case CLEAR_ALL_SETTINGS:
        return tablesOptionsInitialState;

      default:
        return state;
    }
  };

  var setFilters = function setFilters(_ref2) {
    var filters = _ref2.filters;
    return {
      type: SET_FILTERS,
      filters: filters
    };
  };

  var openFilter = function openFilter() {
    return {
      type: OPEN_FILTER
    };
  };

  var closeFilter = function closeFilter() {
    return {
      type: CLOSE_FILTER
    };
  };

  var clearFilter = function clearFilter() {
    return {
      type: CLEAR_FILTER
    };
  };

  var applyFilter = function applyFilter() {
    return {
      type: APPLY_FILTER
    };
  };

  var clearAllSettings = function clearAllSettings() {
    return {
      type: CLEAR_ALL_SETTINGS
    };
  };

  return {
    tableOptionsReducer: tableOptionsReducer,
    setFilters: setFilters,
    openFilter: openFilter,
    closeFilter: closeFilter,
    clearFilter: clearFilter,
    applyFilter: applyFilter,
    clearAllSettings: clearAllSettings
  };
};

var _default = tableReducer;
exports.default = _default;
