// ark-tools/lib/reducers/menuReducer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var R = _interopRequireWildcard(require("ramda"));

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

var menuReducer = function menuReducer(_ref) {
  var key = _ref.key;
  var TOGGLE_MENU_STATE = "".concat(key, "/TOGGLE_MENU_STATE");
  var TOGGLE_PINNED_STATE = "".concat(key, "/TOGGLE_PINNED_STATE");
  var menuReducerState = {
    open: false,
    pinned: false
  };

  var menuStateReducer = function menuStateReducer() {
    var state = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : menuReducerState;
    var action = arguments.length > 1 ? arguments[1] : undefined;

    switch (action.type) {
      case TOGGLE_MENU_STATE:
        return R.evolve({
          open: R.not
        }, state);

      case TOGGLE_PINNED_STATE:
        return R.evolve({
          pinned: R.not
        }, state);

      default:
        return menuReducerState;
    }
  };

  var toggleMenu = function toggleMenu() {
    return {
      type: TOGGLE_MENU_STATE
    };
  };

  var togglePinned = function togglePinned() {
    return {
      type: TOGGLE_PINNED_STATE
    };
  };

  return {
    menuStateReducer: menuStateReducer,
    toggleMenu: toggleMenu,
    togglePinned: togglePinned
  };
};

var _default = menuReducer;
exports.default = _default;
