// ark-tools/lib/reducers/tableFilterReducer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var tableReducer = function tableReducer(_ref) {
  var key = _ref.key,
      _ref$data = _ref.data,
      data = _ref$data === void 0 ? [] : _ref$data,
      _ref$pageSize = _ref.pageSize,
      pageSize = _ref$pageSize === void 0 ? 50 : _ref$pageSize,
      _ref$pageSizeList = _ref.pageSizeList,
      pageSizeList = _ref$pageSizeList === void 0 ? [50, 100, 250] : _ref$pageSizeList,
      _ref$pageNo = _ref.pageNo,
      pageNo = _ref$pageNo === void 0 ? 1 : _ref$pageNo,
      _ref$sortColumn = _ref.sortColumn,
      sortColumn = _ref$sortColumn === void 0 ? null : _ref$sortColumn,
      _ref$columnFilters = _ref.columnFilters,
      columnFilters = _ref$columnFilters === void 0 ? [] : _ref$columnFilters,
      _ref$tableCount = _ref.tableCount,
      tableCount = _ref$tableCount === void 0 ? 0 : _ref$tableCount,
      _ref$filterState = _ref.filterState,
      filterState = _ref$filterState === void 0 ? false : _ref$filterState,
      _ref$filters = _ref.filters,
      filters = _ref$filters === void 0 ? [] : _ref$filters,
      _ref$filterOrig = _ref.filterOrig,
      filterOrig = _ref$filterOrig === void 0 ? [] : _ref$filterOrig,
      _ref$dateRange = _ref.dateRange,
      dateRange = _ref$dateRange === void 0 ? null : _ref$dateRange,
      _ref$datePicker = _ref.datePicker,
      datePicker = _ref$datePicker === void 0 ? null : _ref$datePicker;
  var SET_DATA = "".concat(key, "/SET_DATA");
  var CLEAR_DATA = "".concat(key, "/CLEAR_DATA");
  var SET_PAGE_SIZE = "".concat(key, "/SET_PAGE_SIZE");
  var SET_PAGE_NO = "".concat(key, "/SET_PAGE_NO");
  var SET_SORT_COLUMN = "".concat(key, "/SET_SORT_COLUMN");
  var CLEAR_SORT_COLUMN = "".concat(key, "/CLEAR_SORT_COLUMN");
  var SET_COLUMN_FILTERS = "".concat(key, "/SET_COLUMN_FILTERS");
  var CLEAR_COLUMN_FILTERS = "".concat(key, "/CLEAR_COLUMN_FILTERS");
  var SET_TABLE_COUNT = "".concat(key, "/SET_TABLE_COUNT");
  var RESET_TABLE_SETTINGS = "".concat(key, "/RESET_TABLE_SETTINGS");
  var SET_FILTERS = "".concat(key, "/SET_FILTERS");
  var OPEN_FILTER = "".concat(key, "/OPEN_FILTER");
  var CLOSE_FILTER = "".concat(key, "/CLOSE_FILTER");
  var CLEAR_FILTER = "".concat(key, "/CLEAR_FILTER");
  var APPLY_FILTER = "".concat(key, "/APPLY_FILTER");
  var CLEAR_ALL_SETTINGS = "".concat(key, "/CLEAR_ALL_SETTINGS");
  var SET_DATE_RANGE = "".concat(key, "/SET_DATE_RANGE");
  var CLEAR_DATE_RANGE = "".concat(key, "/CLEAR_DATE_RANGE");
  var SET_DATE_PICKER = "".concat(key, "/SET_DATE_PICKER");
  var CLEAR_DATE_PICKER = "".concat(key, "/CLEAR_DATE_PICKER");
  var tablesOptionsInitialState = {
    data: data,
    pageSize: pageSize,
    pageSizeList: pageSizeList,
    pageNo: pageNo,
    sortColumn: sortColumn,
    columnFilters: columnFilters,
    tableCount: tableCount,
    filterState: filterState,
    filters: filters,
    filterOrig: filterOrig,
    dateRange: dateRange,
    datePicker: datePicker
  };

  var tableOptionsReducer = function tableOptionsReducer() {
    var state = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : tablesOptionsInitialState;
    var action = arguments.length > 1 ? arguments[1] : undefined;

    switch (action.type) {
      case SET_DATA:
        return _objectSpread({}, state, {
          data: action.data
        });

      case CLEAR_DATA:
        return _objectSpread({}, state, {
          data: tablesOptionsInitialState.data
        });

      case SET_PAGE_SIZE:
        return _objectSpread({}, state, {
          pageSize: action.pageSize
        });

      case SET_PAGE_NO:
        return _objectSpread({}, state, {
          pageNo: action.pageNo
        });

      case SET_SORT_COLUMN:
        return _objectSpread({}, state, {
          sortColumn: sortColumn
        });

      case CLEAR_SORT_COLUMN:
        return _objectSpread({}, state, {
          sortColumn: tablesOptionsInitialState.sortColumn
        });

      case SET_COLUMN_FILTERS:
        return _objectSpread({}, state, {
          columnFilters: action.columnFilters
        });

      case CLEAR_COLUMN_FILTERS:
        return _objectSpread({}, state, {
          columnFilters: tablesOptionsInitialState.columnFilters
        });

      case SET_TABLE_COUNT:
        return _objectSpread({}, state, {
          tableCount: action.tableCount
        });

      case RESET_TABLE_SETTINGS:
        return _objectSpread({}, state, {
          pageSize: tablesOptionsInitialState.pageSize,
          pageNo: tablesOptionsInitialState.pageNo,
          sortColumn: tablesOptionsInitialState.sortColumn,
          columnFilters: tablesOptionsInitialState.columnFilters,
          tableCount: tablesOptionsInitialState.tableCount
        });

      case SET_FILTERS:
        return _objectSpread({}, state, {
          filters: action.filters
        });

      case OPEN_FILTER:
        return _objectSpread({}, state, {
          filterState: true,
          filterOrig: state.filters
        });

      case CLOSE_FILTER:
        return _objectSpread({}, state, {
          filterState: false,
          filters: state.filterOrig
        });

      case CLEAR_FILTER:
        return _objectSpread({}, state, {
          pageNo: tablesOptionsInitialState.pageNo,
          filterState: tablesOptionsInitialState.filterState,
          filters: tablesOptionsInitialState.filters
        });

      case APPLY_FILTER:
        return _objectSpread({}, state, {
          pageNo: tablesOptionsInitialState.pageNo,
          filterState: false,
          filterOrig: state.filters
        });

      case SET_DATE_RANGE:
        return _objectSpread({}, state, {
          pageNo: tablesOptionsInitialState.pageNo,
          dateRange: {
            startDate: action.startDate,
            endDate: action.endDate
          }
        });

      case CLEAR_DATE_RANGE:
        return _objectSpread({}, state, {
          pageNo: tablesOptionsInitialState.pageNo,
          dateRange: {
            startDate: null,
            endDate: null
          }
        });

      case SET_DATE_PICKER:
        return _objectSpread({}, state, {
          pageNo: tablesOptionsInitialState.pageNo,
          datePicker: action.date
        });

      case CLEAR_DATE_PICKER:
        return _objectSpread({}, state, {
          pageNo: tablesOptionsInitialState.pageNo,
          datePicker: null
        });

      case CLEAR_ALL_SETTINGS:
        return tablesOptionsInitialState;

      default:
        return state;
    }
  };

  var setData = function setData(_ref2) {
    var data = _ref2.data;
    return {
      type: SET_DATA,
      data: data
    };
  };

  var clearData = function clearData() {
    return {
      type: CLEAR_DATA
    };
  };

  var setPageSize = function setPageSize(_ref3) {
    var pageSize = _ref3.pageSize;
    return {
      type: SET_PAGE_SIZE,
      pageSize: pageSize
    };
  };

  var setPageNo = function setPageNo(_ref4) {
    var pageNo = _ref4.pageNo;
    return {
      type: SET_PAGE_NO,
      pageNo: pageNo
    };
  };

  var setSortColumn = function setSortColumn(_ref5) {
    var col = _ref5.col,
        dir = _ref5.dir;
    return {
      type: SET_SORT_COLUMN,
      SET_SORT_COLUMN: "".concat(col, " ").concat(dir)
    };
  };

  var setTableCount = function setTableCount(_ref6) {
    var tableCount = _ref6.tableCount;
    return {
      type: SET_TABLE_COUNT,
      tableCount: tableCount
    };
  };

  var setColumnFilters = function setColumnFilters(_ref7) {
    var date = _ref7.date;
    return {
      type: SET_COLUMN_FILTERS,
      date: date
    };
  };

  var clearColumnFilters = function clearColumnFilters() {
    return {
      type: CLEAR_COLUMN_FILTERS
    };
  };

  var resetTableSettings = function resetTableSettings() {
    return {
      type: RESET_TABLE_SETTINGS
    };
  };

  var setFilters = function setFilters(_ref8) {
    var filters = _ref8.filters;
    return {
      type: SET_FILTERS,
      filters: filters
    };
  };

  var openFilter = function openFilter() {
    return {
      type: OPEN_FILTER
    };
  };

  var closeFilter = function closeFilter() {
    return {
      type: CLOSE_FILTER
    };
  };

  var clearFilter = function clearFilter() {
    return {
      type: CLEAR_FILTER
    };
  };

  var applyFilter = function applyFilter() {
    return {
      type: APPLY_FILTER
    };
  };

  var clearAllSettings = function clearAllSettings() {
    return {
      type: CLEAR_ALL_SETTINGS
    };
  };

  var setDateRange = function setDateRange(_ref9) {
    var startDate = _ref9.startDate,
        endDate = _ref9.endDate;
    return {
      type: SET_DATE_RANGE,
      startDate: startDate,
      endDate: endDate
    };
  };

  var clearDateRange = function clearDateRange() {
    return {
      type: CLEAR_DATE_RANGE
    };
  };

  var setDatePicker = function setDatePicker(_ref10) {
    var date = _ref10.date;
    return {
      type: SET_DATE_PICKER,
      date: date
    };
  };

  var clearDatePicker = function clearDatePicker() {
    return {
      type: CLEAR_DATE_PICKER
    };
  };

  return {
    tableOptionsReducer: tableOptionsReducer,
    setData: setData,
    clearData: clearData,
    setPageSize: setPageSize,
    setPageNo: setPageNo,
    setSortColumn: setSortColumn,
    setColumnFilters: setColumnFilters,
    clearColumnFilters: clearColumnFilters,
    setTableCount: setTableCount,
    resetTableSettings: resetTableSettings,
    setFilters: setFilters,
    openFilter: openFilter,
    closeFilter: closeFilter,
    clearFilter: clearFilter,
    applyFilter: applyFilter,
    clearAllSettings: clearAllSettings,
    setDateRange: setDateRange,
    clearDateRange: clearDateRange,
    setDatePicker: setDatePicker,
    clearDatePicker: clearDatePicker
  };
};

var _default = tableReducer;
exports.default = _default;
