// ark-tools/lib/components/datePicker/datePicker.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

require("react-dates/initialize");

var R = _interopRequireWildcard(require("ramda"));

var _recompose = require("recompose");

var _reactDates = require("react-dates");

require("react-dates/lib/css/_datepicker.css");

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var OnlyLabelName = function OnlyLabelName(_ref) {
  var labelName = _ref.labelName;
  return _react.default.createElement("label", {
    htmlFor: "select"
  }, _react.default.createElement("b", null, labelName));
};

var RequiredLabelName = function RequiredLabelName(_ref2) {
  var labelName = _ref2.labelName,
      requiredFieldText = _ref2.requiredFieldText;
  return _react.default.createElement("label", {
    htmlFor: "select"
  }, _react.default.createElement("div", {
    className: "requiredLabel"
  }, _react.default.createElement("b", null, labelName), _react.default.createElement("div", {
    className: "required"
  }, requiredFieldText)));
};

var DatePicker = function DatePicker(_ref3) {
  var labelName = _ref3.labelName,
      requiredFieldText = _ref3.requiredFieldText,
      value = _ref3.value,
      _ref3$id = _ref3.id,
      id = _ref3$id === void 0 ? "arkDatePicker" : _ref3$id,
      changeData = _ref3.changeData,
      focusedInput = _ref3.focusedInput,
      setFocusedInput = _ref3.setFocusedInput,
      format = _ref3.format,
      _ref3$disabled = _ref3.disabled,
      disabled = _ref3$disabled === void 0 ? false : _ref3$disabled,
      _ref3$readOnly = _ref3.readOnly,
      readOnly = _ref3$readOnly === void 0 ? false : _ref3$readOnly,
      _ref3$placeHolder = _ref3.placeHolder,
      placeHolder = _ref3$placeHolder === void 0 ? "Choose Date" : _ref3$placeHolder,
      _ref3$withPortal = _ref3.withPortal,
      withPortal = _ref3$withPortal === void 0 ? false : _ref3$withPortal,
      _ref3$numberOfMonths = _ref3.numberOfMonths,
      numberOfMonths = _ref3$numberOfMonths === void 0 ? 1 : _ref3$numberOfMonths;
  return _react.default.createElement(_react.default.Fragment, null, !R.isEmpty(labelName) && R.isEmpty(requiredFieldText) ? _react.default.createElement(OnlyLabelName, {
    labelName: labelName
  }) : null, !R.isEmpty(labelName) && !R.isEmpty(requiredFieldText) ? _react.default.createElement(RequiredLabelName, {
    labelName: labelName,
    requiredFieldText: requiredFieldText
  }) : null, _react.default.createElement(_reactDates.SingleDatePicker, {
    id: id,
    date: value,
    focused: focusedInput,
    onDateChange: changeData,
    onFocusChange: setFocusedInput,
    showClearDate: true,
    small: true,
    disabled: disabled,
    readOnly: readOnly,
    hideKeyboardShortcutsPanel: true,
    displayFormat: format,
    numberOfMonths: numberOfMonths,
    withPortal: withPortal,
    placeholder: placeHolder
  }));
};

var _default = (0, _recompose.compose)((0, _recompose.withState)("focusedInput", "updateFocusedInput", false), (0, _recompose.withHandlers)({
  changeData: function changeData(_ref4) {
    var update = _ref4.update;
    return function (val) {
      return update(val);
    };
  },
  setFocusedInput: function setFocusedInput(props) {
    return function (_ref5) {
      var focused = _ref5.focused;
      return props.updateFocusedInput(focused);
    };
  }
}))(DatePicker);

exports.default = _default;
