// ark-tools/lib/components/dropdown/dropdown.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactSelect = _interopRequireWildcard(require("react-select"));

var R = _interopRequireWildcard(require("ramda"));

var _reactCustomScrollbars = _interopRequireDefault(require("react-custom-scrollbars"));

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var sortByNameCaseInsensitive = R.sortBy(R.compose(R.toLower, R.prop("value")));

var OnlyLabelName = function OnlyLabelName(_ref) {
  var labelName = _ref.labelName;
  return _react.default.createElement("label", {
    htmlFor: "select"
  }, _react.default.createElement("b", null, labelName));
};

var RequiredLabelName = function RequiredLabelName(_ref2) {
  var labelName = _ref2.labelName,
      requiredFieldText = _ref2.requiredFieldText;
  return _react.default.createElement("label", {
    htmlFor: "select"
  }, _react.default.createElement("div", {
    className: "requiredLabel"
  }, _react.default.createElement("b", null, labelName), _react.default.createElement("div", {
    className: "required"
  }, requiredFieldText)));
};

var sortOptionsFunc = function sortOptionsFunc(_ref3) {
  var options = _ref3.options,
      sortOptions = _ref3.sortOptions;
  return sortOptions ? sortByNameCaseInsensitive(options) : options;
};

var ScrollbarsMenu = function ScrollbarsMenu(props) {
  return _react.default.createElement(_reactCustomScrollbars.default, {
    autoHeight: true,
    universal: true,
    renderTrackHorizontal: function renderTrackHorizontal(p) {
      return _react.default.createElement("div", _extends({}, p, {
        style: {
          display: "none"
        },
        className: "track-horizontal"
      }));
    },
    renderThumbHorizontal: function renderThumbHorizontal(vals) {
      return _react.default.createElement("div", _extends({}, vals, {
        className: "thumb-horizontal"
      }));
    },
    renderThumbVertical: function renderThumbVertical(vals) {
      return _react.default.createElement("div", _extends({}, vals, {
        className: "thumb-vertical"
      }));
    }
  }, props.children);
};

var Menu = function Menu(props) {
  return _react.default.createElement(_reactSelect.components.Menu, props, props.menuShouldScrollIntoView ? _react.default.createElement(ScrollbarsMenu, props) : _react.default.createElement("div", {
    className: "arkSelectBody"
  }, props.children));
};

var Dropdown = function Dropdown(_ref4) {
  var labelName = _ref4.labelName,
      selectName = _ref4.selectName,
      change = _ref4.change,
      value = _ref4.value,
      options = _ref4.options,
      _ref4$sortOptions = _ref4.sortOptions,
      sortOptions = _ref4$sortOptions === void 0 ? false : _ref4$sortOptions,
      _ref4$isMulti = _ref4.isMulti,
      isMulti = _ref4$isMulti === void 0 ? false : _ref4$isMulti,
      _ref4$customScrollBar = _ref4.customScrollBars,
      customScrollBars = _ref4$customScrollBar === void 0 ? false : _ref4$customScrollBar,
      _ref4$isDisabled = _ref4.isDisabled,
      isDisabled = _ref4$isDisabled === void 0 ? false : _ref4$isDisabled,
      _ref4$isSearchable = _ref4.isSearchable,
      isSearchable = _ref4$isSearchable === void 0 ? false : _ref4$isSearchable,
      _ref4$isClearable = _ref4.isClearable,
      isClearable = _ref4$isClearable === void 0 ? false : _ref4$isClearable,
      _ref4$closeMenuOnSele = _ref4.closeMenuOnSelect,
      closeMenuOnSelect = _ref4$closeMenuOnSele === void 0 ? true : _ref4$closeMenuOnSele,
      _ref4$noOptionsMessag = _ref4.noOptionsMessage,
      _noOptionsMessage = _ref4$noOptionsMessag === void 0 ? "No Options" : _ref4$noOptionsMessag,
      requiredFieldText = _ref4.requiredFieldText,
      _ref4$placeholder = _ref4.placeholder,
      placeholder = _ref4$placeholder === void 0 ? "Select..." : _ref4$placeholder;

  return _react.default.createElement(_react.default.Fragment, null, !R.isEmpty(labelName) && R.isEmpty(requiredFieldText) ? _react.default.createElement(OnlyLabelName, {
    labelName: labelName
  }) : null, !R.isEmpty(labelName) && !R.isEmpty(requiredFieldText) ? _react.default.createElement(RequiredLabelName, {
    labelName: labelName,
    requiredFieldText: requiredFieldText
  }) : null, _react.default.createElement(_reactSelect.default, {
    isMulti: isMulti,
    className: "arkSelect",
    classNamePrefix: "arkSelect",
    name: selectName,
    value: value,
    onChange: change,
    options: sortOptionsFunc({
      options: options,
      sortOptions: sortOptions
    }),
    isDisabled: isDisabled,
    isSearchable: isSearchable,
    isClearable: isClearable,
    menuShouldScrollIntoView: customScrollBars,
    closeMenuOnSelect: closeMenuOnSelect,
    noOptionsMessage: function noOptionsMessage() {
      return _noOptionsMessage;
    },
    components: {
      Menu: Menu
    },
    placeholder: placeholder
  }));
};

var _default = Dropdown;
exports.default = _default;
