// ark-tools/lib/components/modal/flterModal.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ConfigModal = function ConfigModal(_ref) {
  var open = _ref.open,
      id = _ref.id,
      title = _ref.title,
      headerClass = _ref.headerClass,
      bodyClass = _ref.bodyClass,
      footerClass = _ref.footerClass,
      closeModal = _ref.closeModal,
      modalBody = _ref.modalBody,
      clear = _ref.clear,
      filter = _ref.filter,
      disableFilter = _ref.disableFilter;
  return _react.default.createElement("div", {
    className: "modal ".concat(open ? "appear" : ""),
    id: id,
    role: "dialog",
    "aria-hidden": "true"
  }, _react.default.createElement("div", {
    className: open ? "modalOverlay" : "",
    onClick: closeModal
  }), _react.default.createElement("div", {
    className: "modal-dialog",
    role: "document"
  }, _react.default.createElement("div", {
    className: "modal-content"
  }, _react.default.createElement("div", {
    className: "modal-header ".concat(headerClass)
  }, _react.default.createElement("h5", {
    className: "modal-title"
  }, title)), _react.default.createElement("div", {
    className: "modal-body ".concat(bodyClass)
  }, modalBody), _react.default.createElement("div", {
    className: "modal-footer ".concat(footerClass)
  }, _react.default.createElement("button", {
    type: "button",
    className: "btn btn-default filter cancelFilter",
    onClick: closeModal
  }, "Cancel"), _react.default.createElement("button", {
    type: "button",
    className: "btn btn-danger filter clearFilter",
    onClick: clear
  }, "Clear"), _react.default.createElement("button", {
    type: "button",
    className: "btn btn-danger filter filterSubmit",
    onClick: filter,
    disabled: disableFilter
  }, "Filter")))));
};

var _default = ConfigModal;
exports.default = _default;
